prediction/src/preprocess/improved_lung_segmentation.py

```diff
@@ -0,0 +1,444 @@
+    coords = [(np.mean(np.where(bool_slice), axis=1), i)
+              for i, bool_slice in enumerate(bronchial)]
+    el = min(coords, key=lambda x: sum((x[0] - center) ** 2))
+    return bronchial[el[1]], ct_slices[el[1]], levels[el[1]]
+
+
+def select_seeds(bronchial, ct_clice):
+    """Selection initial pixel.
+
+    Args:
+        bronchial: a label of the pixels with the bronchi;
+        ct_clice:
+
+    Returns:
+        min pixel in area of bronchi
+
+    """
+    return (ct_clice * bronchial) == ct_clice[bronchial].min()
+
+
+def extract_seeds(patient):
+    """ Extraction initial region.
+
+    Starting at the top, the first 25 slices are examined for
+    suitable regions.
+
+    Args:
+        patient (lnp.ndarray[bool]): a label of CT;
+
+    Returns:
+        np.ndarray[bool]: seeds - a initial region.
+
+    """
+    bronchials = list()
+    bronch_cts = list()
+    levels = list()
+    for i in range(25):
+        bronchial = extract_bronchial(patient[i])
+        if bronchial is not None:
+            bronchials.append(bronchial)
+            bronch_cts.append(patient[i])
+            levels.append(i)
+
+    for i in range(-25, 0, 1):
+        bronchial = extract_bronchial(patient[i])
+        if bronchial is not None:
+            bronchials.append(bronchial)
+            bronch_cts.append(patient[i])
+            levels.append(i)
+
+    bronchial, ct_slice, level = select_bronchial(bronchials,
+                                                  bronch_cts,
+                                                  levels)
+
+    seeds = np.zeros(patient.shape)
+    seeds[level] = select_seeds(bronchial, ct_slice)
+
+    return seeds
+
+
+def growing_bronchis(patient, seeds,
+                     threshold=-950,
+                     step=64,
+                     full_extraction=True):
+    """Extraction of large airways.
+
+    Before the lungs are segmented, the trachea and
+    the bronchi are extracted using the technique of a region growing.
+    To initialize the region growing, the start point is automatically
+    identified on the upper axial scan slices by searching for related
+    areas with the specified properties.
+    When an explosion occurred, the increase in threshold was divided by 2.
+
+    Args:
+        patient: a label of CT;
+        seeds: a initial region (pixel);
+        threshold: an average HU;
+        step: In each iteration, the threshold was initially increased with 64 HU
+        full_extraction: full extraction of large airways.
+
+    Returns:
+        ret - a label of the bronchi at the previous stage.
+        seeds - a label of the bronchi.
+
+    """
+
+    seeds = seeds.astype(np.bool_)
+    seeds = region_growing(patient.copy(), seeds, -5010, threshold)
+    volume = np.count_nonzero(seeds)
+
+    lungs_thresh = skimage.filters.threshold_otsu(patient[patient.shape[0] // 2])
+
+    ret = None
+    while True:
+        labeled = region_growing(patient.copy(), seeds, -5010, threshold + step)
```

Comment:
where does `-1510` come from?

Reply:
-5010 is the minimum value in the Hounsfield units that trachea and bronchi can take